new AddressBookController

gvaughn opened the pull request:
This is the followup to my prior address book work. It:

provides endpoints suitable for a custom address book management UI
adds a Spree:Config preference: automatic_default_address
test coverage

Here's some more background. Sorry I didn't write this up when I originally submitted the PR.

The work goes back to making Address a readonly model and avoiding cloning the Address everywhere. Given that readonly behavior, Address is no longer a resource; it's a value object. I have tried to encapsulate the logic that manages copy-on-write semantics for Address. I strongly did not want every controller that works with every model that has an Address to deal with the logic.

At the UserAddressBook concern level there's an additional constraint that I want to encapsulate in the model layer and not require controllers to deal with -- when a UserAddress is marked as default, any other UserAddress marked as default for that user has to also change (sort order also changes). The set of addresses is the unit of encapsulation. That's the core reason that I wrote this code with the Address _BOOK_ as the conceptual resource and the reason I return the entire list. I'd welcome any high level discussion of it at this level, but that's hard to achieve in a PR format.

gvaughn commented on api/app/controllers/spree/api/address_book_controller.rb:
```diff
@@ -0,0 +1,43 @@
+      private
+
+      def address_book_params
+        # TODO id is in the permitted_attributes file, but is gone by this point
```
I'm looking into this, but I wanted to get the PR started and hope someone could shortcut this. permitted_address_attributes is initialized with an :id but from looking at things in a debugger it's gone by this point. I hope it's not being mutated somewhere.

Also the added :default field is rather specific to this controller. I don't mind it being here much, but I'm open to reconfiguring it elsewhere. Briefly: this controller somewhat flattens the UserAddress join model so that :default is another field of the address json object over the wire

jhawthorn commented on api/app/controllers/spree/api/address_book_controller.rb:
```diff
@@ -0,0 +1,43 @@
+         @user_addresses = current_api_user.user_addresses
+      end
+
+      def update
```
This seems more like a create than an update

gvaughn replied:
It supports adding or updating individual address entries in the overall address book.

grzlus replied:
So name it `create_or_update` and make alias.

gvaughn replied:
If I'm deviating too far from "The Rails Way" I'm happy to have that discussion. I'm not creating an AddressBook in this action. Also, Addresses are readonly models, so they cannot be updated. This felt like the best compromise to me -- to think of the AddressBook as the resource.

jhawthorn commented on api/config/routes.rb:
```diff
@@ -133,6 +133,16 @@
+      member do
+        post :remove
+      end
+    end
```
This seems to go out of the way to avoid being the standard rails rest resource. Is there a reason the standard `index`, `create`, and `destroy` were avoided?

gvaughn replied:
It's because the AddressBook itself is the resource. It took me bit to get my head around it.

grzlus replied:
I don't get it. You don't use normal REST resource, but you pass params there. What exactly you want to do with that code?

grzlus replied:
As I understand relation correctly, then we should have `resources :addresses` that is automativally in User scope. User can manage his addresses.

magnusvk replied:
I see the argument of "the address book itself is the resource." But if it's taking all of us a while to wrap our heads around, then maybe that's just not worth it? It wasn't intuitive to me either. What if we went with scoping it in the user? I.e. `/api/users/34/addresses`, `/api/users/34/addresses/1234`? That would have the added advantage of allowing you to build an admin panel around that API that manages others' addresses (though then we'd have to check for that permission). That seems more intuitive to me.

gvaughn replied:
My biggest concern is encapsulation of knowledge and enforcement of constraints. Since only one address (per user) can be default, an edit to one address may cause changes to another (making one default also changes any prior default address). If we treat it in a more traditional Rails-flavored REST-ish way then the client has to know about that and re-request the entire list. We're spreading business logic out into tiers of the app where it doesn't belong. Can we have a brainstorming session in the morning to find the best compromise?

That said, the admin side of things is a good point. However, since this is currently identify users by the auth token, that's the bigger concern.

jhawthorn commented on api/spec/controllers/spree/api/address_book_controller_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+      it "the first one is default" do
+        api_get :mine
+        first, second = *json_response
```
I don't see a reason that they would guaranteed to be returned in this order (unless I missed a default scope somewhere). Maybe it would be better to test that just one of the two was default

gvaughn replied:
The UserAddress join model has the default scope to guarantee this ordering. I can add a unit test for that too.

jhawthorn replied:
Ah, I missed that this was added :+1:

gvaughn replied:
@jhawthorn Is that a general thumbs up for the PR or just for that one question?

grzlus commented on api/app/controllers/spree/api/address_book_controller.rb:
```diff
@@ -0,0 +1,43 @@
+        address = current_api_user.save_in_address_book(address_params, default_flag)
+        if address.valid?
+          @user_addresses = current_api_user.user_addresses
+          render :mine, status: :ok
```
I think that this is a good place to introduce `303 See Other` HTTP status. I use it in my projects, to minimize amount of duplicated code.

Currently I see that in 3 places code `@user_addresses = current_api_user.user_addresses` is used. You can move it to some method, to avoid duplications, but I prefer solution with redirect. When non-GET action is successfull, then server can respond with `redirect_to action: :mine, code: 303`. This means: Everything was successfull, to get information about current resource see this URL.

I know that this is another request, but browsers follow that redirections automatically.

gvaughn replied:
Can you explain the advantage of an additional round-trip from browser to server? As is, the code is updating an AddressBook and receiving the updated AddressBook in the response.

grzlus replied:
Ok, so first of all Law of Demeter. Your code shoud be simple, do only one thing. In that case responsibility of rendering resource is in other action, so redirect user to correct resource to get it.

Second case could sounds strange, but I know from experience, that some servers has garbage collector enabled between requests. So faster response (with redirection) means faster garbage collection and freeing server resources.

And third - If I want use API without getting resources rendered on every update, I can disable follow redirects, because after all updates I'll pull it in one request. It's also edge case.

gvaughn replied:
We understand Law of Demeter quite differently, but nevertheless, I developed this API at the same time as a UI that uses it (which I hope we can generify and share with the community). In every case we wanted to render the updated AddressBook after every modification. I still can't see a 303 status code and extra round-trip as an improvement. I also checked the solidus code and see no use of a 303 or a :see_other status code, so it's not a convention already in place that this code goes against.

grzlus replied:
As I mentioned in first comment - this is good time to discuss adding 303 to project, because you added new code, where we can use it.

magnusvk replied:
I'm :-1: on the 303 as well; it's another HTTP request we're talking about there and I don't really see for what given that we've extracted the method.

grzlus commented on core/app/models/spree/user_address.rb:
```diff
@@ -4,12 +4,18 @@ class UserAddress < Spree::Base
       )
     end
+
+    scope :all_historical, ->() do
```
Don't left empty parentes in lambdas.
This scope is so small, that should be in one line.

gvaughn replied:
Thanks. I'll fix that while I figure out the failing tests

grzlus commented on core/app/models/spree/user_address.rb:
```diff
@@ -4,12 +4,18 @@ class UserAddress < Spree::Base
 
     validates_uniqueness_of :address_id, scope: :user_id
-    validates_uniqueness_of :user_id, conditions: -> { where(default: true) }, message: :default_address_exists, if: :default?
+    validates_uniqueness_of :user_id, conditions: -> { where(default: true, archived: false) }, message: :default_address_exists, if: :default?
```
This conditions could be moved to 2 scopes: `default` and `not_archived`. This could be reusable in other places.

gvaughn replied:
Thanks. I'll fix that while I figure out the failing tests

gvaughn commented:
I've fixed up the small scale suggestions on the PR and fixed the test failures. I've provided more background context in the description, which I should have done initially. The 303 approach suggested needs more community agreement before doing it.

Can I have a re-review?

BenMorganIO commented:
I'm :-1: on a 303. The only reason I could see someone wanting to use it is if you had to do some redirect from one app to another app. We want things to be fast and adding another HTTP request to things doesn't do that.

grzlus commented:
Ok, I don't insist on 303 redirects. It was just a proposition.

One thing that I'd like to see in controller code, is extract logic for selectiong collection and rendering it to separate method. Currently this is used in 3 places in your controller.

gvaughn commented:
@grzlus For the extraction you're suggesting -- something like?:

```ruby
def render_address_book
  @user_addresses = current_api_user.user_addresses
  render :mine, status: :ok
end
```

BenMorganIO commented:
`:mine` doesn't seem restful. We could nest the address information in and make a model to manage the address book contents?

gvaughn commented:
I don't mind changing the name of the action from `mine` but there is precedent in the routes file. Making a separate model would really just be a proxy for a user, which we currently reference by the auth token in the header.

grzlus commented:
@gvaughn Exactly. You put ogic of rendering in one method, so in future, when you decide to add for example decorator, you change it only once.

gvaughn commented:
@grzlus OK. I have a new commit with that now. Yes, I'm well familiar with the advantages. In this case I was on the fence because this was a 2 minute job.

magnusvk commented on api/spec/controllers/spree/api/address_book_controller_spec.rb:
```diff
@@ -0,0 +1,60 @@
+        api_get :mine
+        first, second = *json_response
+        expect(first["default"]).to be true
+        expect(second["default"]).to_not be true
```
Should this be `.to be false` instead? Or is it just falsey?

gvaughn replied:
"not true" is just where my head was when I wrote it. I might be nil though since the db column is not non-nullable

magnusvk commented on api/spec/controllers/spree/api/address_book_controller_spec.rb:
```diff
@@ -0,0 +1,60 @@
+
+      it "can remove an address" do
+        api_post :remove, address_id: address1.id
+        expect(json_response.length).to eq 1
```
Any reason not to go check it's actually gone in the database?

gvaughn replied:
It's still in the db, but I could check for the 'archived' flag in the db. I wouldn't normally do that in a controller test though. The unit tests cover that behavior.

magnusvk replied:
yeah that's fair; I was thinking the same when I got to the unit test. let's leave it

magnusvk commented on api/spec/controllers/spree/api/address_book_controller_spec.rb:
```diff
@@ -0,0 +1,60 @@
+        updated_params[:firstname] = "Johnny"
+        updated_params[:default] = true
+        api_post :update, address_book: updated_params
+        expect(json_response.first["firstname"]).to eq "Johnny"
```
Same here. My instinct would be to check that the update happened in the database.

gvaughn replied:
As above, I don't normally do that, but I'm not dogmatic about it. If it'll get a thumbsup on the PR, I'll make it so.

gvaughn commented:
New commit added with more traditional routing
//cc @jhawthorn @magnusvk

magnusvk commented:
:+1: from me. @jhawthorn?

jhawthorn commented on api/app/controllers/spree/api/address_books_controller.rb:
```diff
@@ -0,0 +1,44 @@
+    class AddressBooksController < Spree::Api::BaseController
+      # Note: the AddressBook is the resource to think about here, not individual addresses
+
+      before_filter :ensure_api_token
```
I believe that this behaviour exists by default. (specs still pass with it removed)

jhawthorn commented on api/app/views/spree/api/address_books/show.v1.rabl:
```diff
@@ -0,0 +1,5 @@
+node do |user_address|
+  partial("spree/api/addresses/show", object: user_address.address).merge(default: user_address.default)
+end
+
```
extra newline

jhawthorn commented:
other than two nitpicks looks good :+1:

gvaughn commented:
New commit added. I'll get it merged on Monday.
